# Data Mining, an introduction to the Pandas package 
This is a companion notebook for video content presented as part of the Data Mining course at SMU.

In this tutorial we will be looking at a number of different parts of the Pandas package for data analysis, including:
- Data Frames
 - loading data
 - head and tail commands
- Munging
 - indexing operations
 - basic statistics
 - encoding
 - imputation (optional)
- bonus: calling R with magics

## Data Frames in Pandas
Data frames in Pandas are basically like tables of data that you can do some really interesting relational database operations upon. There are many built in methods for aggregation and visualization, but we will cover those next time.+

## Data Frames in R
The data frames in Pandas were designed provide the same data manipulation functionality as data frames within R.  Once you understand the Pandas data frame, you are well on your way to understanding the R data frame.  You can check out the following website for a detailed comparison between data frames using Pandas vs. R:

[Data Frames in Pandas vs. R](http://pandas.pydata.org/pandas-docs/stable/comparison_with_r.html)

First lets load a typical table of data from a csv file. You can download the file from here:
https://archive.ics.uci.edu/ml/datasets/Heart+Disease

Make sure to place it in this directory or adjust the path for the file.
### Reading Data from CSV with Pandas

In [15]:
#Python
# let's print out the first five rows inside a csv file

# NOTE: you may need to change the path to the file, 
#       depending on where you saved the data
with open('data/heart_disease.csv') as fid:
    for idx, row in enumerate(fid):
        print row,
        if idx >= 4:
            break

site,age,is_male,chest_pain,rest_blood_press,cholesterol,high_blood_sugar,rest_ecg,max_heart_rate,exer_angina,ST_depression,Peak_ST_seg,major_vessels,thal,has_heart_disease
cleve,63,1,1,145,233,1,2,150,0,2.3,3,0,6,0
cleve,67,1,4,160,286,0,2,108,1,1.5,2,3,3,2
cleve,67,1,4,120,229,0,2,129,1,2.6,2,2,7,1
cleve,37,1,3,130,250,0,0,187,0,3.5,3,0,3,0


In [16]:
#Pandas
# now let's read in the same data to save it as a dataframe
import pandas as pd

df = pd.read_csv('data/heart_disease.csv') # read in the csv file

In [3]:
#Pandas
# now lets look at the data
df.head()

site  age  is_male  chest_pain rest_blood_press cholesterol  \
0  cleve   63        1           1              145         233   
1  cleve   67        1           4              160         286   
2  cleve   67        1           4              120         229   
3  cleve   37        1           3              130         250   
4  cleve   41        0           2              130         204   

  high_blood_sugar rest_ecg max_heart_rate exer_angina ST_depression  \
0                1        2            150           0           2.3   
1                0        2            108           1           1.5   
2                0        2            129           1           2.6   
3                0        0            187           0           3.5   
4                0        2            172           0           1.4   

  Peak_ST_seg major_vessels thal  has_heart_disease  
0           3             0    6                  0  
1           2             3    3                  2  
2           2             2    7                  1  
3           3             0    3                  0  
4           1             0    3                  0

### Installing the R Kernel for use in iPython Notebook

###### Using Anaconda, it is relatively simple to install the R kernel for iPython Notebook.  This is done using the following command from any console window:

conda install -c r r-essentials

This actually uses anaconda's r channel and searches for the package r-essentials to install.  The r-essentials package includes the IRKernel and over 80 of the most used R packages for data science, including dplyr, shiny, ggplot2, tidyr,caret, nnet, and many others!    

###### To avoid naming conflicts, every R package avaiable  within Anaconda's r channel uses the same name as its corresponding R library, except prefixed with "r-".  For example, if you would normally access a package in R using the following library command:

library(SparkR)

###### You would use the following command to install this package using conda: 

conda install -c r r-SparkR

You will see some examples of this later on!

In [63]:
# now let's read in the same data using R and then save it as a dataframe

# set the working directory (this may come in handy sometimes when using R or R Studio)
#setwd("D:\\SMU\\Larson\DataMiningClass\\2U_DataMining\\Jakes Notebooks\data")  

df = inputData <- read.csv("data/heart_disease.csv",sep = ",", header = T)


In [64]:
# now lets look at the data using R
head(df)

site age is_male chest_pain rest_blood_press cholesterol high_blood_sugar
1 cleve  63       1          1              145         233                1
2 cleve  67       1          4              160         286                0
3 cleve  67       1          4              120         229                0
4 cleve  37       1          3              130         250                0
5 cleve  41       0          2              130         204                0
6 cleve  56       1          2              120         236                0
  rest_ecg max_heart_rate exer_angina ST_depression Peak_ST_seg major_vessels
1        2            150           0           2.3           3             0
2        2            108           1           1.5           2             3
3        2            129           1           2.6           2             2
4        0            187           0           3.5           3             0
5        2            172           0           1.4           1             0
6        0            178           0           0.8           1             0
  thal has_heart_disease
1    6                 0
2    3                 2
3    7                 1
4    3                 0
5    3                 0
6    3                 0

In [4]:
#Pandas
# now let's a get a summary of the variables using Pandas
print df.info()
# we can see that most of the data 
#  is saved as an integer or as a nominal object

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 15 columns):
site                 920 non-null object
age                  920 non-null int64
is_male              920 non-null int64
chest_pain           920 non-null int64
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null int64
dtypes: int64(4), object(11)None


In [65]:
#R
# now let's a get a summary of the variables using R
str(df)
# We can also get additional variable information using summary() 
summary(df)
# we can see that most of the data 
#  is saved as an integer or as a factor 

'data.frame':	920 obs. of  15 variables:
 $ site             : Factor w/ 4 levels "cleve","hungary",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ age              : int  63 67 67 37 41 56 62 57 63 53 ...
 $ is_male          : int  1 1 1 1 0 1 0 0 1 1 ...
 $ chest_pain       : int  1 4 4 3 2 2 4 4 4 4 ...
 $ rest_blood_press : Factor w/ 62 levels "?","0","100",..: 36 45 18 27 27 18 33 18 27 33 ...
 $ cholesterol      : Factor w/ 218 levels "?","0","100",..: 87 140 83 104 58 90 122 193 108 57 ...
 $ high_blood_sugar : Factor w/ 3 levels "?","0","1": 3 2 2 2 2 2 2 2 2 3 ...
 $ rest_ecg         : Factor w/ 4 levels "?","0","1","2": 4 4 4 2 4 2 4 2 4 4 ...
 $ max_heart_rate   : Factor w/ 120 levels "?","100","102",..: 51 9 30 87 73 79 61 64 48 56 ...
 $ exer_angina      : Factor w/ 3 levels "?","0","1": 2 3 3 2 2 2 2 3 2 3 ...
 $ ST_depression    : Factor w/ 54 levels "-0.1","-0.5",..: 35 27 38 45 26 20 46 18 26 42 ...
 $ Peak_ST_seg      : Factor w/ 4 levels "?","1","2","3": 4 3 3 4 2 2 4 2 3 4 ...
 $ ma

      site          age           is_male         chest_pain   rest_blood_press
 cleve  :303   Min.   :28.00   Min.   :0.0000   Min.   :1.00   120    :131     
 hungary:294   1st Qu.:47.00   1st Qu.:1.0000   1st Qu.:3.00   130    :115     
 swiss  :123   Median :54.00   Median :1.0000   Median :4.00   140    :102     
 va     :200   Mean   :53.51   Mean   :0.7891   Mean   :3.25   ?      : 59     
               3rd Qu.:60.00   3rd Qu.:1.0000   3rd Qu.:4.00   110    : 59     
               Max.   :77.00   Max.   :1.0000   Max.   :4.00   150    : 56     
                                                               (Other):398     
  cholesterol  high_blood_sugar rest_ecg max_heart_rate exer_angina
 0      :172   ?: 90            ?:  2    ?      : 55    ?: 55      
 ?      : 30   0:692            0:551    150    : 43    0:528      
 220    : 10   1:138            1:179    140    : 41    1:337      
 254    : 10                    2:188    120    : 35               
 204    :  9        

This data has been read into working memory and is known as a DataFrame.

### Reading Data from SQLite3 with Pandas
We can also connect to a sqlite3 database using the built in sqlite3 package that ships with python. This data will be read into working memory and is known as a DataFrame. 

In [5]:
#Pandas
# but csv files are not the only thing we can work with
# what if the data was actually in a sqlite database?
del df
import sqlite3

con = sqlite3.connect('data/heart_disease_sql') # again this file is in the same directory
df = pd.read_sql('SELECT * FROM heart_disease', con)  # the table name is heart_disease
df.head()

site age is_male chest_pain rest_blood_press cholesterol high_blood_sugar  \
0  cleve  63       1          1              145         233                1   
1  cleve  67       1          4              160         286                0   
2  cleve  67       1          4              120         229                0   
3  cleve  37       1          3              130         250                0   
4  cleve  41       0          2              130         204                0   

  rest_ecg max_heart_rate exer_angina ST_depression Peak_ST_seg major_vessels  \
0        2            150           0           2.3           3             0   
1        2            108           1           1.5           2             3   
2        2            129           1           2.6           2             2   
3        0            187           0           3.5           3             0   
4        2            172           0           1.4           1             0   

  thal has_heart_disease  
0    6                 0  
1    3                 2  
2    7                 1  
3    3                 0  
4    3                 0

In [6]:
df.info()
# notice now, however, that the data types are all objects!

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 15 columns):
site                 920 non-null object
age                  920 non-null object
is_male              920 non-null object
chest_pain           920 non-null object
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null object
dtypes: object(15)

### Installing R Packages for use in R, R Studio, and iPython Notebook

###### When using R or R Studio, the following commands will first install and then import any available package using R:

install.packages("RSQLite") # packages only need to be installed one time!

library("RSQLite") # the library() command imports the package from that point on. 

###### When using iPython Notebook however, R packages must installed in a slightly different manner.  While many packages are included with r-essentials, you will eventually run into a package such as RSQLite which is missing.  When this happens, you can install them using the following conda syntax from any command prompt:

conda install -c r r-RSQLite

###### Notice that "r-" is appended to the R package name.  Here is an example of what this would look like in windows:

<img src="condaRpackageInstall.png">


### Reading Data from SQLite3 with R using the RSQLite Package
We can also connect to a sqlite3 database using the built in RSQLite package. 

In [66]:
#R
# but csv files are not the only thing we can work with
# what if the data was actually in a sqlite database?

#Only install one time!
#conda install -c r r-RSQLite

library("RSQLite")

# connect to the sqlite file
con <- dbConnect(RSQLite::SQLite(),dbname="data/heart_disease_sql")
df <- dbGetQuery(con,'SELECT * FROM heart_disease')
head(df)

site age is_male chest_pain rest_blood_press cholesterol high_blood_sugar
1 cleve  63       1          1              145         233                1
2 cleve  67       1          4              160         286                0
3 cleve  67       1          4              120         229                0
4 cleve  37       1          3              130         250                0
5 cleve  41       0          2              130         204                0
6 cleve  56       1          2              120         236                0
  rest_ecg max_heart_rate exer_angina ST_depression Peak_ST_seg major_vessels
1        2            150           0           2.3           3             0
2        2            108           1           1.5           2             3
3        2            129           1           2.6           2             2
4        0            187           0           3.5           3             0
5        2            172           0           1.4           1             0
6        0            178           0           0.8           1             0
  thal has_heart_disease
1    6                 0
2    3                 2
3    7                 1
4    3                 0
5    3                 0
6    3                 0

In [67]:
#Inspect the fields using R
str(df)
# notice now, however, that the data types are all chr!

'data.frame':	920 obs. of  15 variables:
 $ site             : chr  "cleve" "cleve" "cleve" "cleve" ...
 $ age              : chr  "63" "67" "67" "37" ...
 $ is_male          : chr  "1" "1" "1" "1" ...
 $ chest_pain       : chr  "1" "4" "4" "3" ...
 $ rest_blood_press : chr  "145" "160" "120" "130" ...
 $ cholesterol      : chr  "233" "286" "229" "250" ...
 $ high_blood_sugar : chr  "1" "0" "0" "0" ...
 $ rest_ecg         : chr  "2" "2" "2" "0" ...
 $ max_heart_rate   : chr  "150" "108" "129" "187" ...
 $ exer_angina      : chr  "0" "1" "1" "0" ...
 $ ST_depression    : chr  "2.3" "1.5" "2.6" "3.5" ...
 $ Peak_ST_seg      : chr  "3" "2" "2" "3" ...
 $ major_vessels    : chr  "0" "3" "2" "0" ...
 $ thal             : chr  "6" "3" "7" "3" ...
 $ has_heart_disease: chr  "0" "2" "1" "0" ...


 ### Working with DataFrames using Pandas and R
 We can index into a DataFrame in a number of ways:

In [7]:
#Pandas
# the variable names are embedded into the structure
print df.age
print df['age'] # but can also be accessed using strings

0     63
1     67
2     67
3     37
4     41
5     56
6     62
7     57
8     63
9     53
10    57
11    56
12    56
13    44
14    52
...
905    41
906    43
907    44
908    47
909    47
910    49
911    49
912    50
913    50
914    52
915    52
916    54
917    56
918    58
919    65
Name: age, Length: 920, dtype: object
0     63
1     67
2     67
3     37
4     41
5     56
6     62
7     57
8     63
9     53
10    57
11    56
12    56
13    44
14    52
...
905    41
906    43
907    44
908    47
909    47
910    49
911    49
912    50
913    50
914    52
915    52
916    54
917    56
918    58
919    65
Name: age, Length: 920, dtype: object


In [68]:
#===========================================================
#R
# the variable names are embedded into the structure but accessed using a $ character
df$age
df['age'] # but can also be accessed using strings

[1] "63" "67" "67" "37" "41" "56" "62" "57" "63" "53" "57" "56" "56" "44" "52"
 [16] "57" "48" "54" "48" "49" "64" "58" "58" "58" "60" "50" "58" "66" "43" "40"
 [31] "69" "60" "64" "59" "44" "42" "43" "57" "55" "61" "65" "40" "71" "59" "61"
 [46] "58" "51" "50" "65" "53" "41" "65" "44" "44" "60" "54" "50" "41" "54" "51"
 [61] "51" "46" "58" "54" "54" "60" "60" "54" "59" "46" "65" "67" "62" "65" "44"
 [76] "65" "60" "51" "48" "58" "45" "53" "39" "68" "52" "44" "47" "53" "53" "51"
 [91] "66" "62" "62" "44" "63" "52" "59" "60" "52" "48" "45" "34" "57" "71" "49"
[106] "54" "59" "57" "61" "39" "61" "56" "52" "43" "62" "41" "58" "35" "63" "65"
[121] "48" "63" "51" "55" "65" "45" "56" "54" "44" "62" "54" "51" "29" "51" "43"
[136] "55" "70" "62" "35" "51" "59" "59" "52" "64" "58" "47" "57" "41" "45" "60"
[151] "52" "42" "67" "55" "64" "70" "51" "58" "60" "68" "46" "77" "54" "58" "48"
[166] "57" "52" "54" "35" "45" "70" "53" "59" "62" "64" "57" "52" "56" "43" "53"
[181] "48" "56" "42" "59" "60" "63" "42" "66" "54" "69" "50" "51" "43" "62" "68"
[196] "67" "69" "45" "50" "59" "50" "64" "57" "64" "43" "45" "58" "50" "55" "62"
[211] "37" "38" "41" "66" "52" "56" "46" "46" "64" "59" "41" "54" "39" "53" "63"
[226] "34" "47" "67" "54" "66" "52" "55" "49" "74" "54" "54" "56" "46" "49" "42"
[241] "41" "41" "49" "61" "60" "67" "58" "47" "52" "62" "57" "58" "64" "51" "43"
[256] "42" "67" "76" "70" "57" "44" "58" "60" "44" "61" "42" "52" "59" "40" "42"
[271] "61" "66" "46" "71" "59" "64" "66" "39" "57" "58" "57" "47" "55" "35" "61"
[286] "58" "58" "58" "56" "56" "67" "55" "44" "63" "63" "41" "59" "57" "45" "68"
[301] "57" "57" "38" "32" "34" "35" "36" "38" "38" "38" "38" "38" "38" "40" "41"
[316] "42" "42" "43" "43" "43" "45" "46" "46" "47" "47" "47" "47" "48" "50" "50"
[331] "50" "50" "51" "51" "51" "51" "51" "51" "51" "52" "52" "52" "52" "53" "53"
[346] "53" "53" "53" "53" "53" "53" "54" "54" "54" "55" "55" "55" "55" "56" "56"
[361] "56" "56" "56" "56" "56" "56" "57" "57" "57" "57" "57" "57" "58" "58" "58"
[376] "59" "59" "59" "59" "59" "60" "60" "60" "60" "60" "60" "61" "61" "61" "61"
[391] "61" "61" "61" "61" "61" "62" "62" "62" "62" "62" "62" "62" "63" "63" "63"
[406] "63" "63" "64" "64" "64" "65" "65" "65" "65" "66" "66" "67" "68" "68" "69"
[421] "69" "70" "70" "72" "73" "74" "63" "44" "60" "55" "66" "66" "65" "60" "60"
[436] "60" "56" "59" "62" "63" "57" "62" "63" "46" "63" "60" "58" "64" "63" "74"
[451] "52" "69" "51" "60" "56" "55" "54" "77" "63" "55" "52" "64" "60" "60" "58"
[466] "59" "61" "40" "61" "41" "57" "63" "59" "51" "59" "42" "55" "63" "62" "56"
[481] "53" "68" "53" "60" "62" "59" "51" "61" "57" "56" "58" "69" "67" "58" "65"
[496] "63" "55" "57" "65" "54" "72" "75" "49" "51" "60" "64" "58" "61" "67" "62"
[511] "65" "63" "69" "51" "62" "55" "75" "40" "67" "58" "60" "63" "35" "62" "43"
[526] "63" "68" "65" "48" "63" "64" "61" "50" "59" "55" "45" "65" "61" "49" "72"
[541] "50" "64" "55" "63" "59" "56" "62" "74" "54" "57" "62" "76" "54" "70" "61"
[556] "48" "48" "61" "66" "68" "55" "62" "71" "74" "53" "58" "75" "56" "58" "64"
[571] "54" "54" "59" "55" "57" "61" "41" "71" "38" "55" "56" "69" "64" "72" "69"
[586] "56" "62" "67" "57" "69" "51" "48" "69" "69" "64" "57" "53" "37" "67" "74"
[601] "63" "58" "61" "64" "58" "60" "57" "55" "55" "56" "57" "61" "61" "58" "74"
[616] "68" "51" "62" "53" "62" "46" "54" "62" "55" "58" "62" "28" "29" "29" "30"
[631] "31" "32" "32" "32" "33" "34" "34" "34" "35" "35" "35" "35" "36" "36" "36"
[646] "36" "37" "37" "37" "37" "37" "37" "37" "38" "38" "38" "39" "39" "39" "39"
[661] "39" "39" "39" "39" "39" "39" "40" "40" "40" "40" "40" "41" "41" "41" "41"
[676] "41" "41" "41" "42" "42" "42" "42" "42" "42" "42" "43" "43" "43" "43" "43"
[691] "43" "43" "43" "44" "44" "44" "44" "45" "45" "45" "45" "45" "45" "45" "46"
[706] "46" "46" "46" "46" "46" "46" "47" "47" "47" "47" "47" "48" "48" "48" "48"
[721] "48" "48" "48" "48" "48" "48" "48" "49" "49" "49" "49" "49" "49" "49" "49"
[736] "50" "50" "50" "50" "50" "5

age
1    63
2    67
3    67
4    37
5    41
6    56
7    62
8    57
9    63
10   53
11   57
12   56
13   56
14   44
15   52
16   57
17   48
18   54
19   48
20   49
21   64
22   58
23   58
24   58
25   60
26   50
27   58
28   66
29   43
30   40
31   69
32   60
33   64
34   59
35   44
36   42
37   43
38   57
39   55
40   61
41   65
42   40
43   71
44   59
45   61
46   58
47   51
48   50
49   65
50   53
51   41
52   65
53   44
54   44
55   60
56   54
57   50
58   41
59   54
60   51
61   51
62   46
63   58
64   54
65   54
66   60
67   60
68   54
69   59
70   46
71   65
72   67
73   62
74   65
75   44
76   65
77   60
78   51
79   48
80   58
81   45
82   53
83   39
84   68
85   52
86   44
87   47
88   53
89   53
90   51
91   66
92   62
93   62
94   44
95   63
96   52
97   59
98   60
99   52
100  48
101  45
102  34
103  57
104  71
105  49
106  54
107  59
108  57
109  61
110  39
111  61
112  56
113  52
114  43
115  62
116  41
117  58
118  35
119  63
120  65
121  48
122  63
123  51
124  55
125  65
126  45
127  56
128  54
129  44
130  62
131  54
132  51
133  29
134  51
135  43
136  55
137  70
138  62
139  35
140  51
141  59
142  59
143  52
144  64
145  58
146  47
147  57
148  41
149  45
150  60
151  52
152  42
153  67
154  55
155  64
156  70
157  51
158  58
159  60
160  68
161  46
162  77
163  54
164  58
165  48
166  57
167  52
168  54
169  35
170  45
171  70
172  53
173  59
174  62
175  64
176  57
177  52
178  56
179  43
180  53
181  48
182  56
183  42
184  59
185  60
186  63
187  42
188  66
189  54
190  69
191  50
192  51
193  43
194  62
195  68
196  67
197  69
198  45
199  50
200  59
201  50
202  64
203  57
204  64
205  43
206  45
207  58
208  50
209  55
210  62
211  37
212  38
213  41
214  66
215  52
216  56
217  46
218  46
219  64
220  59
221  41
222  54
223  39
224  53
225  63
226  34
227  47
228  67
229  54
230  66
231  52
232  55
233  49
234  74
235  54
236  54
237  56
238  46
239  49
240  42
241  41
242  41
243  49
244  61
245  60
246  67
247  58
248  47
249  52
250  62
251  57
252  58
253  64
254  51
255  43
256  42
257  67
258  76
259  70
260  57
261  44
262  58
263  60
264  44
265  61
266  42
267  52
268  59
269  40
270  42
271  61
272  66
273  46
274  71
275  59
276  64
277  66
278  39
279  57
280  58
281  57
282  47
283  55
284  35
285  61
286  58
287  58
288  58
289  56
290  56
291  67
292  55
293  44
294  63
295  63
296  41
297  59
298  57
299  45
300  68
301  57
302  57
303  38
304  32
305  34
306  35
307  36
308  38
309  38
310  38
311  38
312  38
313  38
314  40
315  41
316  42
317  42
318  43
319  43
320  43
321  45
322  46
323  46
324  47
325  47
326  47
327  47
328  48
329  50
330  50
331  50
332  50
333  51
334  51
335  51
336  51
337  51
338  51
339  51
340  52
341  52
342  52
343  52
344  53
345  53
346  53
347  53
348  53
349  53
350  53
351  53
352  54
353  54
354  54
355  55
356  55
357  55
358  55
359  56
360  56
361  56
362  56
363  56
364  56
365  56
366  56
367  57
368  57
369  57
370  57
371  57
372  57
373  58
374  58
375  58
376  59
377  59
378  59
379  59
380  59
381  60
382  60
383  60
384  60
385  60
386  60
387  61
388  61
389  61
390  61
391  61
392  61
393  61
394  61
395  61
396  62
397  62
398  62
399  62
400  62
401  62
402  62
403  63
404  63
405  63
406  63
407  63
408  64
409  64
410  64
411  65
412  65
413  65
414  65
415  66
416  66
417  67
418  68
419  68
420  69
421  69
422  70
423  70
424  72
425  73
426  74
427  63
428  44
429  60
430  55
431  66
432  66
433  65
434  60
435  60
436  60
437  56
438  59
439  62
440  63
441  57
442  62
443  63
444  46
445  63
446  60
447  58
448  64
449  63
450  74
451  52
452  69
453  51
454  60
455  56
456  55
457  54
458  77
459  63
460  55
461  52
462  64
463  60
464  60
465  58
466  59
467  61
468  40
469  61
470  41
471  57
472  63
473  59
474  51
475  59
476  42
477  55
478  63
479  62
480  56
481  53
482  68
483  53
484  60
485  62
486  59
487  51
488  61
489  57
490  56
491  58
492  69
493  67
494  58
495  65
496  63
497  55
498  57
499  65
500 

In [8]:
#Pandas
print df.chest_pain.min(), df.chest_pain.max(), df.chest_pain.mean()

1 4 inf


In [69]:
#R
paste(min(df$chest_pain), max(df$chest_pain), mean(df$chest_pain), sep=" ")

Warning message:
In mean.default(df$chest_pain): argument is not numeric or logical: returning NA

[1] "1 4 NA"

In [9]:
#Pandas
# lets get rid of the 'site' variable
if 'site' in df:
    del df['site']

print df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null object
is_male              920 non-null object
chest_pain           920 non-null object
rest_blood_press     920 non-null object
cholesterol          920 non-null object
high_blood_sugar     920 non-null object
rest_ecg             920 non-null object
max_heart_rate       920 non-null object
exer_angina          920 non-null object
ST_depression        920 non-null object
Peak_ST_seg          920 non-null object
major_vessels        920 non-null object
thal                 920 non-null object
has_heart_disease    920 non-null object
dtypes: object(14)None


In [70]:
#R
# lets get rid of the 'site' variable using R
#df[ SUBSET ROWS HERE , SUBSET COLUMNS HERE]
#See the follwing URL for a million different alternate examples:
#    http://stackoverflow.com/questions/4605206/drop-data-frame-columns-by-name
df <- df[, !(colnames(df) %in% c("site"))] 

#do the same thing using the column index, CAREFUL, IF you run this twice, age will be gone as well!
#This is litterally saying select all columns, except the column at index 1
#df <- df[, -1] 

str(df)

'data.frame':	920 obs. of  14 variables:
 $ age              : chr  "63" "67" "67" "37" ...
 $ is_male          : chr  "1" "1" "1" "1" ...
 $ chest_pain       : chr  "1" "4" "4" "3" ...
 $ rest_blood_press : chr  "145" "160" "120" "130" ...
 $ cholesterol      : chr  "233" "286" "229" "250" ...
 $ high_blood_sugar : chr  "1" "0" "0" "0" ...
 $ rest_ecg         : chr  "2" "2" "2" "0" ...
 $ max_heart_rate   : chr  "150" "108" "129" "187" ...
 $ exer_angina      : chr  "0" "1" "1" "0" ...
 $ ST_depression    : chr  "2.3" "1.5" "2.6" "3.5" ...
 $ Peak_ST_seg      : chr  "3" "2" "2" "3" ...
 $ major_vessels    : chr  "0" "3" "2" "0" ...
 $ thal             : chr  "6" "3" "7" "3" ...
 $ has_heart_disease: chr  "0" "2" "1" "0" ...


In [10]:
#Pandas
# Notice that all of the data is stored as a non-null object
# That's not good. It means we need to change those data types
# in order to encode the variables properly. Right now Pandas
# thinks all of our variables are nominal!

import numpy as np
# replace '?' with -1, we will deal with missing values later
df = df.replace(to_replace='?',value=-999) 

# let's start by first changing the numeric values to be floats
continuous_features = ['rest_blood_press', 'cholesterol', 
                       'max_heart_rate', 'ST_depression']

# and the oridnal values to be integers
ordinal_features = ['age','major_vessels','chest_pain',
                    'rest_ecg','Peak_ST_seg','thal','has_heart_disease']

# we won't touch these variables, keep them as categorical
categ_features = ['is_male','high_blood_sugar','exer_angina'];

# use the "astype" function to change the variable type
df[continuous_features] = df[continuous_features].astype(np.float64)
df[ordinal_features] = df[ordinal_features].astype(np.int64)

df.info() # now our data looks better!!

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null int64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null int64
major_vessels        920 non-null int64
thal                 920 non-null int64
has_heart_disease    920 non-null int64
dtypes: float64(4), int64(7), object(3)

In [71]:
#===========================================================
#R
# Notice that all of the data is stored as character vectors.
# That's not good. It means we need to change those data types
# in order to encode the variables properly. Right now R
# thinks all of our variables are nominal!

# replace '?' with -999, we will deal with missing values later
df[df == '?'] <- -999

# let's start by first changing the numeric values to be floats
continuous_features = c('rest_blood_press', 'cholesterol', 
                       'max_heart_rate', 'ST_depression')

# and the oridnal values to be integers
ordinal_features = c('age','major_vessels','chest_pain',
                    'rest_ecg','Peak_ST_seg','thal','has_heart_disease')

# we won't touch these variables, keep them as categorical
categ_features = c('is_male','high_blood_sugar','exer_angina')

# use the sapply function to change the variable type
df[ , continuous_features] <- lapply(df[,continuous_features],as.numeric)
df[ , ordinal_features] <- lapply(df[,ordinal_features],as.numeric)
df[ , categ_features] <- lapply(df[,categ_features],as.factor)

str(df) # now our data looks better!!

'data.frame':	920 obs. of  14 variables:
 $ age              : num  63 67 67 37 41 56 62 57 63 53 ...
 $ is_male          : Factor w/ 2 levels "0","1": 2 2 2 2 1 2 1 1 2 2 ...
 $ chest_pain       : num  1 4 4 3 2 2 4 4 4 4 ...
 $ rest_blood_press : num  145 160 120 130 130 120 140 120 130 140 ...
 $ cholesterol      : num  233 286 229 250 204 236 268 354 254 203 ...
 $ high_blood_sugar : Factor w/ 3 levels "-999","0","1": 3 2 2 2 2 2 2 2 2 3 ...
 $ rest_ecg         : num  2 2 2 0 2 0 2 0 2 2 ...
 $ max_heart_rate   : num  150 108 129 187 172 178 160 163 147 155 ...
 $ exer_angina      : Factor w/ 3 levels "-999","0","1": 2 3 3 2 2 2 2 3 2 3 ...
 $ ST_depression    : num  2.3 1.5 2.6 3.5 1.4 0.8 3.6 0.6 1.4 3.1 ...
 $ Peak_ST_seg      : num  3 2 2 3 1 1 3 1 2 3 ...
 $ major_vessels    : num  0 3 2 0 0 0 2 0 1 0 ...
 $ thal             : num  6 3 7 3 3 3 3 3 7 7 ...
 $ has_heart_disease: num  0 2 1 0 0 0 3 0 2 1 ...


In [11]:
#Pandas
df.head()

age is_male  chest_pain  rest_blood_press  cholesterol high_blood_sugar  \
0   63       1           1               145          233                1   
1   67       1           4               160          286                0   
2   67       1           4               120          229                0   
3   37       1           3               130          250                0   
4   41       0           2               130          204                0   

   rest_ecg  max_heart_rate exer_angina  ST_depression  Peak_ST_seg  \
0         2             150           0            2.3            3   
1         2             108           1            1.5            2   
2         2             129           1            2.6            2   
3         0             187           0            3.5            3   
4         2             172           0            1.4            1   

   major_vessels  thal  has_heart_disease  
0              0     6                  0  
1              3     3                  2  
2              2     7                  1  
3              0     3                  0  
4              0     3                  0

In [72]:
#R
head(df)

age is_male chest_pain rest_blood_press cholesterol high_blood_sugar rest_ecg
1  63       1          1              145         233                1        2
2  67       1          4              160         286                0        2
3  67       1          4              120         229                0        2
4  37       1          3              130         250                0        0
5  41       0          2              130         204                0        2
6  56       1          2              120         236                0        0
  max_heart_rate exer_angina ST_depression Peak_ST_seg major_vessels thal
1            150           0           2.3           3             0    6
2            108           1           1.5           2             3    3
3            129           1           2.6           2             2    7
4            187           0           3.5           3             0    3
5            172           0           1.4           1             0    3
6            178           0           0.8           1             0    3
  has_heart_disease
1                 0
2                 2
3                 1
4                 0
5                 0
6                 0

Let's get summary of all attributes in the frame

In [12]:
#Pandas
df.describe() # will get summary of continuous or the nominals

age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  920.000000  920.000000        920.000000   920.000000  920.000000   
mean    53.510870    3.250000        123.594565   192.604348    0.601087   
std      9.424685    0.930969         37.484705   114.615011    0.808415   
min     28.000000    1.000000         -1.000000    -1.000000   -1.000000   
25%     47.000000    3.000000        120.000000   164.000000    0.000000   
50%     54.000000    4.000000        130.000000   221.000000    0.000000   
75%     60.000000    4.000000        140.000000   267.000000    1.000000   
max     77.000000    4.000000        200.000000   603.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      920.000000     920.000000   920.000000     920.000000  920.000000   
mean       129.263043       0.752174     0.840217      -0.436957    1.871739   
std         41.376773       1.154353     1.403211       0.959656    3.313649   
min         -1.000000      -2.600000    -1.000000      -1.000000   -1.000000   
25%        115.000000       0.000000    -1.000000      -1.000000   -1.000000   
50%        138.000000       0.200000     1.000000      -1.000000   -1.000000   
75%        156.000000       1.500000     2.000000       0.000000    6.000000   
max        202.000000       6.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count         920.000000  
mean            0.995652  
std             1.142693  
min             0.000000  
25%             0.000000  
50%             1.000000  
75%             2.000000  
max             4.000000

In [73]:
#R
summary(df)

      age        is_male   chest_pain   rest_blood_press   cholesterol    
 Min.   :28.00   0:194   Min.   :1.00   Min.   :-999.00   Min.   :-999.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.: 120.00   1st Qu.: 164.0  
 Median :54.00           Median :4.00   Median : 130.00   Median : 221.0  
 Mean   :53.51           Mean   :3.25   Mean   :  59.59   Mean   : 160.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.: 140.00   3rd Qu.: 267.0  
 Max.   :77.00           Max.   :4.00   Max.   : 200.00   Max.   : 603.0  
 high_blood_sugar    rest_ecg        max_heart_rate   exer_angina
 -999: 90         Min.   :-999.000   Min.   :-999.0   -999: 55   
 0   :692         1st Qu.:   0.000   1st Qu.: 115.0   0   :528   
 1   :138         Median :   0.000   Median : 138.0   1   :337   
                  Mean   :  -1.568   Mean   :  69.6              
                  3rd Qu.:   1.000   3rd Qu.: 156.0              
                  Max.   :   2.000   Max.   : 202.0              
 ST_depressio

There are 920 entries in this data frame. Notice that this data frame has a number of missing values denoted by the value -999 (that we changed the '?' value to before). We need to either remove the missing values from the dataset OR we need to fill in with our best guess for those values. Let's first drop all the rows with missing values.

In [13]:
# how many value have the -1 (which we set as the missing values) 
import numpy as np

# let's set those values to NaN, so that Pandas understand they are missing
df = df.replace(to_replace=-999,value=np.nan) # replace -1 with NaN (not a number)
print df.info()
df.describe() # scroll over to see the values

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     861 non-null float64
cholesterol          890 non-null float64
high_blood_sugar     830 non-null object
rest_ecg             918 non-null float64
max_heart_rate       865 non-null float64
exer_angina          865 non-null object
ST_depression        856 non-null float64
Peak_ST_seg          611 non-null float64
major_vessels        309 non-null float64
thal                 434 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)None


age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  920.000000  920.000000        861.000000   890.000000  918.000000   
mean    53.510870    3.250000        132.132404   199.130337    0.604575   
std      9.424685    0.930969         19.066070   110.780810    0.805827   
min     28.000000    1.000000          0.000000     0.000000    0.000000   
25%     47.000000    3.000000        120.000000   175.000000    0.000000   
50%     54.000000    4.000000        130.000000   223.000000    0.000000   
75%     60.000000    4.000000        140.000000   268.000000    1.000000   
max     77.000000    4.000000        200.000000   603.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      865.000000     856.000000   611.000000     309.000000  434.000000   
mean       137.545665       0.883178     1.770867       0.676375    5.087558   
std         25.926276       1.088707     0.619256       0.935653    1.919075   
min         60.000000      -2.600000     1.000000       0.000000    3.000000   
25%        120.000000       0.000000     1.000000       0.000000    3.000000   
50%        140.000000       0.500000     2.000000       0.000000    6.000000   
75%        157.000000       1.500000     2.000000       1.000000    7.000000   
max        202.000000       6.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count         920.000000  
mean            0.995652  
std             1.142693  
min             0.000000  
25%             0.000000  
50%             1.000000  
75%             2.000000  
max             4.000000

In [74]:
#R
# how many values were previously ? (which we set as the missing values) 
length(df[df == '-999'])  #Count the  -999 values

# let's set those values to NA, so that R understand they are missing
df[df == '-999'] <- NA

# how many values were previously -999 (which we set as the missing values) 
sum(is.na(df))

str(df)
summary(df) # scroll over to see the values

[1] 1759

[1] 1759

'data.frame':	920 obs. of  14 variables:
 $ age              : num  63 67 67 37 41 56 62 57 63 53 ...
 $ is_male          : Factor w/ 2 levels "0","1": 2 2 2 2 1 2 1 1 2 2 ...
 $ chest_pain       : num  1 4 4 3 2 2 4 4 4 4 ...
 $ rest_blood_press : num  145 160 120 130 130 120 140 120 130 140 ...
 $ cholesterol      : num  233 286 229 250 204 236 268 354 254 203 ...
 $ high_blood_sugar : Factor w/ 3 levels "-999","0","1": 3 2 2 2 2 2 2 2 2 3 ...
 $ rest_ecg         : num  2 2 2 0 2 0 2 0 2 2 ...
 $ max_heart_rate   : num  150 108 129 187 172 178 160 163 147 155 ...
 $ exer_angina      : Factor w/ 3 levels "-999","0","1": 2 3 3 2 2 2 2 3 2 3 ...
 $ ST_depression    : num  2.3 1.5 2.6 3.5 1.4 0.8 3.6 0.6 1.4 3.1 ...
 $ Peak_ST_seg      : num  3 2 2 3 1 1 3 1 2 3 ...
 $ major_vessels    : num  0 3 2 0 0 0 2 0 1 0 ...
 $ thal             : num  6 3 7 3 3 3 3 3 7 7 ...
 $ has_heart_disease: num  0 2 1 0 0 0 3 0 2 1 ...


      age        is_male   chest_pain   rest_blood_press  cholesterol   
 Min.   :28.00   0:194   Min.   :1.00   Min.   :  0.0    Min.   :  0.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.:120.0    1st Qu.:175.0  
 Median :54.00           Median :4.00   Median :130.0    Median :223.0  
 Mean   :53.51           Mean   :3.25   Mean   :132.1    Mean   :199.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.:140.0    3rd Qu.:268.0  
 Max.   :77.00           Max.   :4.00   Max.   :200.0    Max.   :603.0  
                                        NA's   :59       NA's   :30     
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 -999:  0         Min.   :0.0000   Min.   : 60.0   -999:  0   
 0   :692         1st Qu.:0.0000   1st Qu.:120.0   0   :528   
 1   :138         Median :0.0000   Median :140.0   1   :337   
 NA's: 90         Mean   :0.6046   Mean   :137.5   NA's: 55   
                  3rd Qu.:1.0000   3rd Qu.:157.0              
                  Max.   :2.0000   Max

Wow. Notice how the number of attributes went down in the description function. Looks like we need to impute values. If we drop the rows with missing data, we will be throwing away almost 80% of the data collected. No way!!

### Imputation of NaN values (Optional)

In [14]:
#Pandas
# lets look at some stats of the data
df.median() # only calculates for numeric data

age                   54.0
is_male                1.0
chest_pain             4.0
rest_blood_press     130.0
cholesterol          223.0
high_blood_sugar       0.0
rest_ecg               0.0
max_heart_rate       140.0
exer_angina            0.0
ST_depression          0.5
Peak_ST_seg            2.0
major_vessels          0.0
thal                   6.0
has_heart_disease      1.0
dtype: float64

In [75]:
#R
# lets look at some stats of the data
summary(df) # only calculates for numeric data

      age        is_male   chest_pain   rest_blood_press  cholesterol   
 Min.   :28.00   0:194   Min.   :1.00   Min.   :  0.0    Min.   :  0.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.:120.0    1st Qu.:175.0  
 Median :54.00           Median :4.00   Median :130.0    Median :223.0  
 Mean   :53.51           Mean   :3.25   Mean   :132.1    Mean   :199.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.:140.0    3rd Qu.:268.0  
 Max.   :77.00           Max.   :4.00   Max.   :200.0    Max.   :603.0  
                                        NA's   :59       NA's   :30     
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 -999:  0         Min.   :0.0000   Min.   : 60.0   -999:  0   
 0   :692         1st Qu.:0.0000   1st Qu.:120.0   0   :528   
 1   :138         Median :0.0000   Median :140.0   1   :337   
 NA's: 90         Mean   :0.6046   Mean   :137.5   NA's: 55   
                  3rd Qu.:1.0000   3rd Qu.:157.0              
                  Max.   :2.0000   Max

In [15]:
#Pandas
# the 'fillna' function will take the given series (the output above)
# and fill in the missing values for the columns it has
df_imputed = df.fillna(df.median()) # note that to do this all values must be numeric
df_imputed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null float64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null float64
major_vessels        920 non-null float64
thal                 920 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)

Notice that the object variables are unchanged, but all the numeric/ordinal values have been filled in with the median of the columns. Let's try something (slightly) smarter, and fill in the oridinals with the median and the continuous with the mean.

In [76]:
#R
# the lapply function will take the given series (the output above)
# and fill in the missing values for the columns it has

df_imputed <- df

#Get only the numeric columns in the data frame
numCols <- sapply(df, is.numeric)

#Now perform Simple Mean Imputation on each numeric column
df_imputed[,numCols] <- lapply(df_imputed[,numCols], function(x) { 
  x[is.na(x)] <- mean(x, na.rm = TRUE)
  x
})

summary(df_imputed)

      age        is_male   chest_pain   rest_blood_press  cholesterol   
 Min.   :28.00   0:194   Min.   :1.00   Min.   :  0.0    Min.   :  0.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.:120.0    1st Qu.:177.8  
 Median :54.00           Median :4.00   Median :130.0    Median :221.0  
 Mean   :53.51           Mean   :3.25   Mean   :132.1    Mean   :199.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.:140.0    3rd Qu.:267.0  
 Max.   :77.00           Max.   :4.00   Max.   :200.0    Max.   :603.0  
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 -999:  0         Min.   :0.0000   Min.   : 60.0   -999:  0   
 0   :692         1st Qu.:0.0000   1st Qu.:120.0   0   :528   
 1   :138         Median :0.0000   Median :138.0   1   :337   
 NA's: 90         Mean   :0.6046   Mean   :137.5   NA's: 55   
                  3rd Qu.:1.0000   3rd Qu.:156.0              
                  Max.   :2.0000   Max.   :202.0              
 ST_depression      Peak_ST_seg    major_vessels

Notice all of the NA values are now imputed / replaced with the mean() for each column.

In [16]:
# Pandas
# make  one series for imputing with
series_mean = df[continuous_features].mean()
series_median = df[categ_features+ordinal_features].median()
cat_series = pd.concat((series_median,series_mean))

print cat_series

is_male                1.000000
high_blood_sugar       0.000000
exer_angina            0.000000
age                   54.000000
major_vessels          0.000000
chest_pain             4.000000
rest_ecg               0.000000
Peak_ST_seg            2.000000
thal                   6.000000
has_heart_disease      1.000000
rest_blood_press     132.132404
cholesterol          199.130337
max_heart_rate       137.545665
ST_depression          0.883178
dtype: float64


In [17]:
# Pandas
# now let's impute the numbers a bit differently

df_imputed = df.fillna(value=cat_series)
df_imputed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null int64
is_male              920 non-null object
chest_pain           920 non-null int64
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null float64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null float64
major_vessels        920 non-null float64
thal                 920 non-null float64
has_heart_disease    920 non-null int64
dtypes: float64(8), int64(3), object(3)

In [77]:
#R
# impute with the mean for the continuous features
df_imputed <- df

df_imputed[,continuous_features] <- lapply(df_imputed[,continuous_features], function(x) { 
  x[is.na(x)] <- mean(x, na.rm = TRUE)
  x
})

# impute with the median for the categ_features + ordinal_features
df_imputed[,c(continuous_features, ordinal_features)] <- lapply(df_imputed[,c(continuous_features, ordinal_features)], function(x) { 
  x[is.na(x)] <- median(x)
  x
})

summary(df_imputed)

      age        is_male   chest_pain   rest_blood_press  cholesterol   
 Min.   :28.00   0:194   Min.   :1.00   Min.   :  0.0    Min.   :  0.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.:120.0    1st Qu.:177.8  
 Median :54.00           Median :4.00   Median :130.0    Median :221.0  
 Mean   :53.51           Mean   :3.25   Mean   :132.1    Mean   :199.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.:140.0    3rd Qu.:267.0  
 Max.   :77.00           Max.   :4.00   Max.   :200.0    Max.   :603.0  
                                                                        
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 -999:  0         Min.   :0.0000   Min.   : 60.0   -999:  0   
 0   :692         1st Qu.:0.0000   1st Qu.:120.0   0   :528   
 1   :138         Median :0.0000   Median :138.0   1   :337   
 NA's: 90         Mean   :0.6046   Mean   :137.5   NA's: 55   
                  3rd Qu.:1.0000   3rd Qu.:156.0              
                  Max.   :2.0000   Max

In [18]:
# Pandas
df_imputed[categ_features].describe()

is_male high_blood_sugar exer_angina
count      920              920         920
unique       2                3           3
top          1                0           0
freq       726              692         528

In [78]:
#R
summary(df_imputed[, categ_features])

 is_male high_blood_sugar exer_angina
 0:194   -999:  0         -999:  0   
 1:726   0   :692         0   :528   
         1   :138         1   :337   
         NA's: 90         NA's: 55   

### Indexing logically into Data Frames
Let's now say that we are only interested in the summary of the dataframe when the patient has heart disease. We can achieve this using a few line of code:

In [19]:
#Pandas
df_imputed[df_imputed.has_heart_disease==0].describe()

age  chest_pain  rest_blood_press  cholesterol    rest_ecg  \
count  411.000000  411.000000        411.000000   411.000000  411.000000   
mean    50.547445    2.761557        130.021042   226.575368    0.547445   
std      9.433700    0.903425         16.460208    74.301504    0.805204   
min     28.000000    1.000000         80.000000     0.000000    0.000000   
25%     43.000000    2.000000        120.000000   199.130337    0.000000   
50%     51.000000    3.000000        130.000000   225.000000    0.000000   
75%     57.000000    4.000000        140.000000   266.000000    1.000000   
max     76.000000    4.000000        190.000000   564.000000    2.000000   

       max_heart_rate  ST_depression  Peak_ST_seg  major_vessels        thal  \
count      411.000000     411.000000   411.000000     411.000000  411.000000   
mean       148.252830       0.441963     1.729927       0.111922    5.085158   
std         23.152969       0.704565     0.515662       0.427276    1.510951   
min         69.000000      -1.100000     1.000000       0.000000    3.000000   
25%        135.500000       0.000000     1.000000       0.000000    3.000000   
50%        150.000000       0.000000     2.000000       0.000000    6.000000   
75%        165.000000       0.883178     2.000000       0.000000    6.000000   
max        202.000000       4.200000     3.000000       3.000000    7.000000   

       has_heart_disease  
count                411  
mean                   0  
std                    0  
min                    0  
25%                    0  
50%                    0  
75%                    0  
max                    0

In [79]:
#R
summary(df_imputed[df_imputed$has_heart_disease==0, ])

      age        is_male   chest_pain    rest_blood_press  cholesterol   
 Min.   :28.00   0:144   Min.   :1.000   Min.   : 80      Min.   :  0.0  
 1st Qu.:43.00   1:267   1st Qu.:2.000   1st Qu.:120      1st Qu.:199.1  
 Median :51.00           Median :3.000   Median :130      Median :225.0  
 Mean   :50.55           Mean   :2.762   Mean   :130      Mean   :226.6  
 3rd Qu.:57.00           3rd Qu.:4.000   3rd Qu.:140      3rd Qu.:266.0  
 Max.   :76.00           Max.   :4.000   Max.   :190      Max.   :564.0  
                                                                         
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 -999:  0         Min.   :0.0000   Min.   : 69.0   -999:  0   
 0   :353         1st Qu.:0.0000   1st Qu.:135.5   0   :336   
 1   : 44         Median :0.0000   Median :150.0   1   : 55   
 NA's: 14         Mean   :0.5474   Mean   :148.3   NA's: 20   
                  3rd Qu.:1.0000   3rd Qu.:165.0              
                  Max.   :2.00

In [20]:
#Pandas
# or we can use the extremely useful "groupby" function
df_imputed.groupby(by='has_heart_disease').median()

age  chest_pain  rest_blood_press  cholesterol  rest_ecg  \
has_heart_disease                                                             
0                   51           3        130.000000        225.0         0   
1                   55           4        130.000000        226.0         0   
2                   58           4        132.132404        193.0         0   
3                   60           4        132.132404        212.0         1   
4                   59           4        133.066202        218.5         1   

                   max_heart_rate  ST_depression  Peak_ST_seg  major_vessels  \
has_heart_disease                                                              
0                           150.0           0.00            2              0   
1                           130.0           1.00            2              0   
2                           130.0           1.40            2              0   
3                           122.0           1.00            2              0   
4                           126.5           2.45            2              0   

                   thal  
has_heart_disease        
0                     6  
1                     6  
2                     6  
3                     6  
4                     6

In [80]:
#R
#or we can use the extremely useful "groupby" function
library(dplyr)
numCols <- sapply(df, is.numeric)

df[,numCols] %>%
  group_by(has_heart_disease) %>%
  summarise_each(funs(median))


Source: local data frame [5 x 11]

  has_heart_disease   age chest_pain rest_blood_press cholesterol rest_ecg
              (dbl) (dbl)      (dbl)            (lgl)       (lgl)    (dbl)
1                 0    51          3               NA          NA        0
2                 1    55          4               NA          NA       NA
3                 2    58          4               NA          NA        0
4                 3    60          4               NA          NA        1
5                 4    59          4               NA          NA        1
Variables not shown: max_heart_rate (lgl), ST_depression (lgl), Peak_ST_seg
  (lgl), major_vessels (lgl), thal (lgl)

In [21]:
#Pandas
df_imputed.groupby(by=df_imputed.has_heart_disease>0).mean()

age  chest_pain  rest_blood_press  cholesterol  \
has_heart_disease                                                         
False              50.547445    2.761557        130.021042   226.575368   
True               55.903733    3.644401        133.837257   176.969418   

                   rest_ecg  max_heart_rate  ST_depression  Peak_ST_seg  \
has_heart_disease                                                         
False              0.547445      148.252830       0.441963     1.729927   
True               0.648330      128.899997       1.239443     1.943026   

                   major_vessels      thal  has_heart_disease  
has_heart_disease                                              
False                   0.111922  5.085158           0.000000  
True                    0.320236  5.960707           1.799607

In [22]:
#Pandas
df_imputed.groupby(by=df_imputed.major_vessels>2).mean()

age  chest_pain  rest_blood_press  cholesterol  rest_ecg  \
major_vessels                                                                   
False          53.368889    3.241111        132.026458   197.656567  0.586667   
True           59.900000    3.650000        136.900000   265.450000  1.350000   

               max_heart_rate  ST_depression  Peak_ST_seg  major_vessels  \
major_vessels                                                              
False              137.612235       0.861359     1.847778       0.165556   
True               134.550000       1.865000     1.850000       3.000000   

                   thal  has_heart_disease  
major_vessels                               
False          5.566667           0.966667  
True           5.700000           2.300000

### One Hot Encoding of Categorical Variables

In [23]:
#Pandas
# one hot encoded variables can be created using the get_dummies variable
tmpdf = pd.get_dummies(df_imputed['chest_pain'],prefix='chest')

tmpdf.head()

chest_1  chest_2  chest_3  chest_4
0        1        0        0        0
1        0        0        0        1
2        0        0        0        1
3        0        0        1        0
4        0        1        0        0

In [24]:
#Pandas
#one hot encoding of ALL categorical variables
# there is lot going on in this one line of code, so let's step through it

# pd.concat([*]], axis=1) // this line of code concatenates all the data frames in the [*] list
# [** for col in categ_features] // this steps through each feature in categ_features and 
#                                //   creates a new element in a list based on the output of **
# pd.get_dummies(df_imputed[col],prefix=col) // this creates a one hot encoded dataframe of the variable=col (like code above)

one_hot_df = pd.concat([pd.get_dummies(df_imputed[col],prefix=col) for col in categ_features], axis=1)

one_hot_df.head()

is_male_0  is_male_1  high_blood_sugar_0.0  high_blood_sugar_0  \
0          0          1                     0                   0   
1          0          1                     0                   1   
2          0          1                     0                   1   
3          0          1                     0                   1   
4          1          0                     0                   1   

   high_blood_sugar_1  exer_angina_0.0  exer_angina_0  exer_angina_1  
0                   1                0              1              0  
1                   0                0              0              1  
2                   0                0              0              1  
3                   0                0              1              0  
4                   0                0              1              0

## Calling R from iPython

- Note: you will need R installed on your machine to run these!!

iPython has a lot of interesting "magics" built in. If you use R and have it installed on your machine, then you can write and look at R code directly from iPython cells. R also uses data frames, which we can push data into directly from the Pandas object we are using:

In [25]:
# CONVERT PANDAS DATAFRAME TO R DATA.FRAME
# adapted from: http://tagteam.harvard.edu/hub_feeds/1981/feed_items/196017
# I have better luck with both calls here

%load_ext rmagic
%load_ext rpy2.ipython

df_colnames = df_imputed.columns

/Library/Python/2.7/site-packages/IPython/extensions/rmagic.py:693: UserWarning: The rmagic extension in IPython is deprecated in favour of rpy2.ipython. If available, that will be loaded instead.
http://rpy.sourceforge.net/
  warnings.warn("The rmagic extension in IPython is deprecated in favour of "


In [26]:
df_colnames

Index([u'age', u'is_male', u'chest_pain', u'rest_blood_press', u'cholesterol', u'high_blood_sugar', u'rest_ecg', u'max_heart_rate', u'exer_angina', u'ST_depression', u'Peak_ST_seg', u'major_vessels', u'thal', u'has_heart_disease'], dtype='object')

Now lets take the data frame from pandas and tell Rmagics that we want to have variables available in the R workspace. We use the %%R command to tell iPython that the entire cell is R code. The "-i" tells Rmagics that we want to transfer those variables over to R.

The following code will take the variables df_imputed and df_colnames into the R workspace and test if they are truly saved as R data.frames type variables.

In [27]:
%%R -i df_imputed,df_colnames 

colnames(df_imputed) <- unlist(df_colnames); 
print(is.data.frame(df_imputed))

[1] TRUE


They were data.frames! Great. Let's call an R function on the data.frame.

In [28]:
%%R -i df_imputed 
print(summary(df_imputed))

      age        is_male   chest_pain   rest_blood_press  cholesterol   
 Min.   :28.00   0:194   Min.   :1.00   Min.   :  0.0    Min.   :  0.0  
 1st Qu.:47.00   1:726   1st Qu.:3.00   1st Qu.:120.0    1st Qu.:177.8  
 Median :54.00           Median :4.00   Median :130.0    Median :221.0  
 Mean   :53.51           Mean   :3.25   Mean   :132.1    Mean   :199.1  
 3rd Qu.:60.00           3rd Qu.:4.00   3rd Qu.:140.0    3rd Qu.:267.0  
 Max.   :77.00           Max.   :4.00   Max.   :200.0    Max.   :603.0  
 high_blood_sugar    rest_ecg      max_heart_rate  exer_angina
 0  :692          Min.   :0.0000   Min.   : 60.0   0  :528    
 0.0: 90          1st Qu.:0.0000   1st Qu.:120.0   0.0: 55    
 1  :138          Median :0.0000   Median :138.0   1  :337    
                  Mean   :0.6033   Mean   :137.5              
                  3rd Qu.:1.0000   3rd Qu.:156.0              
                  Max.   :2.0000   Max.   :202.0              
 ST_depression      Peak_ST_seg    major_vessels

So we are able to call some R and get console output, now let's make some changes to the data.fram in R and print the result back in python.

In [29]:
print 'original:', df_imputed.age.head()

# give df_imputed, then multiply it by to in R
# the %R command tells iPython its just one line of R code
%R -i df_imputed df_imputed$age <- df_imputed$age*2

# now we are back in python, did it change?
print 'after manipulation in R:', df_imputed.age.head()

original: 0    63
1    67
2    67
3    37
4    41
Name: age, dtype: int64
after manipulation in R: 0    63
1    67
2    67
3    37
4    41
Name: age, dtype: int64


Well, it looks like the data was not synchronized... So instead let's setup an output variable for the DataFrame that we send into R. `-i df_imputed` means that we are sending in the DataFrame as an R data.frame. `-o df_imputed` means we are also getting the same variable and copying it back to the python workspace.

In [30]:
print 'original:', df_imputed.age.head() 

# This is the same code as before, but now with an output variable
%R -i df_imputed -o df_imputed  df_imputed$age <- df_imputed$age*2
# you can place the above on any line to make sure that the data stays
# synchronized between pandas and python
print 'after manipulation in R:', df_imputed.age.head()

original: 0    63
1    67
2    67
3    37
4    41
Name: age, dtype: int64
after manipulation in R: 0    126
1    134
2    134
3     74
4     82
Name: age, dtype: float64


Awesome. So now we can send DataFrames into R, manipulate them, and get them back into the python workspace. Is this memory hogging? Yes. Is it really useful for when you want to connect and work with different parts of R? You betcha.

In [31]:
# We can also just go and get new variables from R and 
# have them spit them back out for us
# here I am sending in df_imputed and getting back a data frame
# created in R
%R -i df_imputed -o df_from_R df_from_R <- df_imputed

# notice that the only differebce is that the integers are 32 bits
df_from_R.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 920 entries, 0 to 919
Data columns (total 14 columns):
age                  920 non-null float64
is_male              920 non-null object
chest_pain           920 non-null int32
rest_blood_press     920 non-null float64
cholesterol          920 non-null float64
high_blood_sugar     920 non-null object
rest_ecg             920 non-null float64
max_heart_rate       920 non-null float64
exer_angina          920 non-null object
ST_depression        920 non-null float64
Peak_ST_seg          920 non-null float64
major_vessels        920 non-null float64
thal                 920 non-null float64
has_heart_disease    920 non-null int32
dtypes: float64(9), int32(2), object(3)

That's it. Use this as a reference sheet for Pandas, some basic imputation, and calling R code. Thanks!